File: DDD DSL/DDD DSL Design War Semantics of How Many/Code/Java/src/fayelab/ddd/trianglecounter/original/TriangleCounter.java
package fayelab.ddd.trianglecounter.original;

import java.util.List;
import java.util.stream.Stream;

import static java.util.Arrays.asList;
import static java.util.stream.Collectors.toList;
import static fayelab.ddd.trianglecounter.original.SetOp.subset;
import static fayelab.ddd.trianglecounter.original.Combinator.combinate;

public class TriangleCounter
{
    public static long count(String pointsDesc, String...lineDescs)
    {
        return count(parsePoints(pointsDesc), parseLines(lineDescs));
    }
    
    static List<Character> parsePoints(String pointsDesc)
    {
        return pointsDesc.chars().mapToObj(p -> (char)p).collect(toList());
    }
    
    static List<List<Character>> parseLines(String...lineDescs)
    {
        return Stream.of(lineDescs).map(TriangleCounter::parsePoints).collect(toList());
    }
    
    static boolean belong(List<List<Character>> lines, Character...points)
    {
        return lines.stream().anyMatch(line -> subset(asList(points), line));
    }
    
    static boolean connected(char p1, char p2, List<List<Character>> lines)
    {
        return belong(lines, p1, p2);
    }
    
    static boolean onALine(char p1, char p2, char p3, List<List<Character>> lines)
    {
        return belong(lines, p1, p2, p3);
    }
    
    static boolean triangle(char p1, char p2, char p3, List<List<Character>> lines)
    {
        return connected(p1, p2, lines) 
            && connected(p1, p3, lines)
            && connected(p2, p3, lines)
            && (!onALine(p1, p2, p3, lines));
    }
    
    static long count(List<Character> points, List<List<Character>> lines)
    {
        return combinate(points, 3).stream()
                                   .filter(triplePoints -> triangle(triplePoints.get(0), triplePoints.get(1),
                                                                    triplePoints.get(2), lines))
                                   .count();
    }
}

File: DDD DSL/DDD DSL Design War Semantics of How Many/Code/Python/trianglecounter.py
from setop import belong
from combinator import combinate

def connected(p1, p2, lines):
    return belong([p1, p2], lines)

def on_a_line(p1, p2, p3, lines):
    return belong([p1, p2, p3], lines)

def triangle(p1, p2, p3, lines):
    return connected(p1, p2, lines) and \
           connected(p2, p3, lines) and \
           connected(p1, p3, lines) and \
           (not on_a_line(p1, p2, p3, lines))

def inner_count(points, lines):
    return len([triplepoints for triplepoints in combinate(points, 3)
                if triangle(triplepoints[0], triplepoints[1], triplepoints[2], lines)])

def parse_points(points):
    return [point for point in points]

def parse_lines(lines):
    return [parse_points(line) for line in lines]

def count(points, lines):
    return inner_count(parse_points(points), parse_lines(lines))